### PR TITLE
Adding main script

### PR DESCRIPTION
**Pull request by yujin-kimmm**

### CODE REVIEW

**Review by richa-namballa — CHANGES_REQUESTED** [review]
* Make the script executable from the command line with arguments (`argparse`)
* Put the main code in a function called `main()`
* All functions should be at the top of the script
* Add comments to your code
* The save path of the final WAV file should be configurable with the command line arguments
* For the demucs code, you should check if a gpu is available, if not run on cpu
* The sample rate should be a constant
* If the script becomes too long, I would break it into modules (separate scripts)

**Comment by yujin-kimmm** [comment]
Code test completed.

File: README.md [hunk 1]
```diff
@@ -1 +1,32 @@
...
+## Getting Started
+
+### Dependencies
+
+* python 3.10
+
+### Executing program
+
+```
+python --live path/to/live/audio.wav --recorded path/to/recorded/audio.wav --output_dir path/to/output/directory
```
Comment (richa-namballa) [inline README.md]: Add line breaks to make this easier to read:

```
python --live path/to/live/audio.wav \
       --recorded path/to/recorded/audio.wav \
       --output_dir path/to/output/directory
```

File: main.py [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+    accompaniments = accompaniments.detach().cpu().numpy().T
+    
+    # Convert to mono
+    if len(vocals.shape) != 1:
+        vocals = np.mean(vocals, axis=1)
+    
+    if len(accompaniments.shape) != 1:
+        accompaniments = np.mean(accompaniments, axis=1)
+    
+    return vocals, accompaniments, sr
```
Comment (richa-namballa) [inline main.py]: I may go one step further to make the code more modular and put the separation function in a separate script `separate.py` so that `main.py` only has the main code to parse arguments and run the tool.

File: main.py [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+    
+if __name__ == '__main__':
+
+    # gpu check
+    if torch.cuda.is_available():
+        DEVICE = torch.device("cuda")
+        print("GPU detected: using CUDA.")
+    elif torch.backends.mps.is_available():
+        DEVICE = torch.device("mps")
+        print("MPS detected: using MPS backend.")
```
Comment (richa-namballa) [inline main.py]: This will cause an error in lines 41-42 if you use an MPS backend, I would either remove this clause or add MPS support to the demucs script.

File: main.py [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+    
+    # Apply separation
+    waveform = waveform.unsqueeze(0)
+    sources = apply_model(model, waveform, split=True, overlap=0.25, progress=True)[0]
+
+    # Map sources to names
+    sources_dict = dict(zip(model.sources, sources))
+
+    # Combine drums, bass, and other → accompaniment
+    accompaniments = sources_dict["drums"] + sources_dict["bass"] + sources_dict["other"]
```
Comment (richa-namballa) [inline main.py]: There should be a vocals only option in demucs, no?

File: main.py [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+        vocals = np.mean(vocals, axis=1)
+    
+    if len(accompaniments.shape) != 1:
+        accompaniments = np.mean(accompaniments, axis=1)
+    
+    return vocals, accompaniments, sr
+
+
+def main():
+    parser = argparse.ArgumentParser(description='Live Vocal Extraction.')
```
Comment (richa-namballa) [inline main.py]: Write a more descriptive description 😂

File: main.py [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+    Live_vox, Live_inst, sr_1 = separate(args.live, DEVICE)
+    print("Separating Live audio done.")
+    
+    print("Separating Recorded audio start.")
+    Rec_vox, Rec_inst, sr_2 = separate(args.recorded, DEVICE)
+    print("Separating Recorded audio done.")
+    
+    print("Separating vocals...")
+    
+    SAMPLE_RATE = 44100
```
Comment (richa-namballa) [inline main.py]: Usually, you put constants at the top of a file (after imports) or store them in a separate file to allow for easy configuration.

File: main.py [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+        DEVICE = torch.device("cuda")
+        print("GPU detected: using CUDA.")
+    elif torch.backends.mps.is_available():
+        DEVICE = torch.device("mps")
+        print("MPS detected: using MPS backend.")
+    else:
+        DEVICE = None
+        print("GPU not found. Defaulting to CPU.")
+
+    main()
```
Comment (richa-namballa) [inline main.py]: Add a blank line at the end of every file (this is just good Python practice)

File: scale.py [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+
+    alpha_0 = 0.5
+    
+    lsq_sol = least_squares(residuals, alpha_0,
+                            args=(y1_frames[max_corr_frame, :],
+                                  y2_frames[max_corr_frame, :]))
+    return lsq_sol.x
+
+def residuals(alpha, y1, y2):
+    return (alpha * y1) - y2
```
Comment (richa-namballa) [inline scale.py]: Move this function above `compute_scale_factor` (good practice) and add comments to your code 🙂

File: shift.py [hunk 9]
```diff
@@ -0,0 +1,79 @@
+import numpy as np
+from scipy.stats import mode
+from utils import frame_signal, rms
+
```
Comment (richa-namballa) [inline shift.py:4]: Use two blank lines after imports (formatting thing)

File: utils.py [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+def rms(x):
+    return np.sqrt(np.mean(x ** 2))
+
+def check_sample_rate(x, y, sr, target_sr):
+    if sr != target_sr:
+        num_samples_x = int(x.shape[-1] * target_sr / sr)
+        x = resample(x, num=num_samples_x, axis=-1)
+        num_samples_y = int(y.shape[-1] * target_sr / sr)
+        y = resample(y, num=num_samples_y, axis=-1)
+        sr = target_sr
```
Comment (richa-namballa) [inline utils.py]: Why don't you use librosa's resampling function?

**Review by richa-namballa — CHANGES_REQUESTED** [review]
See comments for small changes. Otherwise, looks good!

File: main.py (hunk 4 above)
Comment (yujin-kimmm) [reply main.py]: They have two stems (vocals only) option in demucs, but that is doing the same thing as I did here, but by command line with arguments. They originally divide into each source -> combine accompaniments and return it. So this code is doing the same thing as demucs two stems, but not in command line. Please correct me if I'm wrong!🙂

File: utils.py (hunk 10 above)
Comment (yujin-kimmm) [reply utils.py]: I changed to librosa resampling

File: main.py (hunk 4 above)
Comment (richa-namballa) [reply main.py]: Ah you're right, I forgot that the option is only available in the CLI. Thanks for checking!

**Review by richa-namballa — APPROVED** [review]
